src/System.Device.Gpio/System/Device/Gpio/GpioController.cs

```diff
@@ -96,6 +96,7 @@ public void OpenPin(int pinNumber)
 
         OpenPinCore(pinNumber);
         _openPins.TryAdd(pinNumber, null);
+        return new GpioPin(pinNumber, this);
```

Comment:
As discussed, it might be valuable to have the driver initialize the pin instead of the controller, in case a subclassed driver wants to return a subclassed Gpio Pin. The way this is today, a subclass of GpioPin can never be created, since OpenPin is not virtual. We should either:

- Have GpioPin be sealed and leave the rest as is.
- Keep GpioPin unsealed but make it abstract, and also change OpenPinCore's return value to be GpioPin. Also changed GpioDriver's OpenPin methods to return GpioPin.

Reply:
I directly pass the driver now and removed the disposal.
I did not touch the driver at all, no need for it to return a GpioPin, the controller keep the control of opening and closing the pins.
I've updated the API design to add the use case of having extender and how pain it is in the current situation vs with a list of GpioPin.

Reply:
Would there ever be a reason for GpioPin to be extended by someone? I kind of thought that there could be cases where we would want to let people extend GpioPin, and if that is the case, then you will have to delegate the task of opening the pins to the driver, or else it will never be possible to create them. For example, imagine I create a Ft422GpioDriver and Ft422GpioPin classes, how can I ever get a Ft422GpioPin object back? This wouldn't be possible even with creating a Ft422GpioController, because OpenPin method that handles opening the pins is not virtual.

If we don't expect people to try to do the scenario I described above, then let's make GpioPin sealed to make our intention clear.

Reply:
Something like this should just work:

```csharp
namepace System.Device.Gpio
{
  public class Ft422GpioPin: GpioPin
  {
     public Ft422GpioPin(int pinNumber, GpioDriver driver): base(pinNumber, driver)
     {
        driver.OpenPin(pinNumber); 
     }
  }
}

// Then usage:
var pin = new Ft422GpioPin(2, new Ft422GpioDriver());
```

Reply:
What is the role of the controller in that case? I think what you are proposing is confusing in two reasons:
- GpioPins now get created in two different ways, depending on if you want a normal pin (in which case you create them via a controller) or a special pin (where you do it via ctor passing in the driver)
- ownership is even more complicated now, who disposes the driver? In our case is the controller, but presumably in your case it is the pin?

My suggestion is to have the same consistent usage no matter if you are a custom pin or not. That can be accomplished simply by also changing the driver's OpenPin to return GPioPin, and that way everything is always created via de controller

Reply:
> My suggestion is to have the same consistent usage no matter if you are a custom pin or not. That can be accomplished simply by also changing the driver's OpenPin to return GPioPin, and that way everything is always created via de controller

I guess you mean by the driver then rather than the controller ;-) That's possible as well.

Note that the example was just to answer your question on how it will be done. To extend a bit, that can as well add a way to close the pin:

```csharp
namepace System.Device.Gpio
{
  public class Ft422GpioPin: GpioPin, IDisposable
  {
     private bool _shouldDispose;
     public Ft422GpioPin(int pinNumber, GpioDriver driver, bool shouldDispose = false): base(pinNumber, driver)
     {
        driver.OpenPin(pinNumber); 
        _should Dispose = shouldDispose;
     }

    public void ClosePin()
    {
      driver.ClosePin(PinNumber);
    }

    public void Dispose()
    {
        if (_shouldDispose)
        {
            driver.Dispose();
        }
    }
  }
}

// Then usage:
var driver = new Ft422GpioDriver();
var pin1 = new Ft422GpioPin(1, driver);
var pin2 = new Ft422GpioPin(2, driver);
// Do whatever 
// Then clean
pin1.Dispose();
pin2.Dispose();
driver.Dispose();
```

Additional thoughts? :-)